tpm/src/emulator.rs

```diff
@@ -0,0 +1,450 @@
+        })?;
+
+        let mut emulator = Self {
+            had_startup_error: false,
+            cmd: None,
+            caps: 0,
+            ctrl_socet: socket,
+            data_fd: -1,
+            established_flag_cached: false,
+            established_flag: false,
+        };
+
+        emulator.prepare_data_fd()?;
+
+        emulator.probe_caps()?;
+        if !emulator.check_caps() {
+            emulator.had_startup_error = true;
+        }
+
+        if !emulator.get_established_flag() {
+            emulator.had_startup_error = true;
+        }
+
+        Ok(emulator)
+    }
+
+    /// Create socketpair, assign one socket/FD as data_fd to ctrl Socket
+    /// The other socket/FD will be assigned to msg_fd, which will be sent to swtpm
+    /// via CmdSetDatafd ctrl command
+    fn prepare_data_fd(&mut self) -> Result<()> {
+        let mut res: PtmResult = 0;
+
+        let mut fds = [-1, -1];
+        unsafe {
+            let ret = libc::socketpair(libc::AF_UNIX, libc::SOCK_STREAM, 0, fds.as_mut_ptr());
+            if ret == -1 {
+                return Err(Error::PrepareDataFd(anyhow!(
+                    "Failed to prepare data fd for tpm emulator. Error Code {:?}",
+                    std::io::Error::last_os_error()
+                )));
+            }
+        }
+        self.ctrl_socet.set_msgfd(fds[1]);
+        debug!("data fd to be configured in swtpm = {:?}", fds[1]);
+        self.ctrlcmd(Commands::CmdSetDatafd, &mut res, 0, mem::size_of::<u32>())?;
+        debug!("data fd in cloud-hypervisor = {:?}", fds[0]);
+        self.data_fd = fds[0];
+        self.ctrl_socet.set_datafd(fds[0]);
+        Ok(())
+    }
+
+    /// Gather TPM Capabilities and cache them in Emulator
+    ///
+    fn probe_caps(&mut self) -> Result<()> {
+        let mut caps: u64 = 0;
+        self.ctrlcmd(
+            Commands::CmdGetCapability,
+            &mut caps,
+            0,
+            mem::size_of::<u64>(),
+        )?;
+        self.caps = caps;
+        Ok(())
+    }
+
+    /// Check if minimum set of capabitlies are supported
+    fn check_caps(&mut self) -> bool {
+        /* min. required capabilities for TPM 2.0*/
+        let caps: PtmCap = PTM_CAP_INIT
+            | PTM_CAP_SHUTDOWN
+            | PTM_CAP_GET_TPMESTABLISHED
+            | PTM_CAP_SET_LOCALITY
+            | PTM_CAP_RESET_TPMESTABLISHED
+            | PTM_CAP_SET_DATAFD
+            | PTM_CAP_STOP
+            | PTM_CAP_SET_BUFFERSIZE;
+
+        if (self.caps & caps) != caps {
+            return false;
+        }
+        true
+    }
+
+    ///
+    /// # Arguments
+    ///
+    /// * `cmd` - Ctrl Command to run
+    /// * `msg` - Optional msg to be sent along with Ctrl Command
+    /// * `msg_len_in` - len of 'msg' in bytes, if passed
+    /// * `msg_len_out` - length of expected output from Ctrl Command in bytes
+    ///
+    fn ctrlcmd(
+        &mut self,
+        cmd: Commands,
+        msg: &mut dyn Ptm,
```

Comment:
Why does this have to be a C style in/out variable. Can't you return separately?

Reply:
I couldn't find a strong reason for doing it this way. The original author wrote it this way and I retained it.
If you are strongly against this, I will take some time to refactor this. If not, I would like to merge this version and revisit it later.